File: backend/services/price_service.py
import os
import httpx
from typing import Dict, Optional
import logging

logger = logging.getLogger(__name__)

class PriceService:
    def __init__(self):
        self.api_key = os.getenv("ONEINCH_API_KEY")
        self.base_url = "https://api.1inch.dev"
        
        # Token addresses for price queries
        self.token_addresses = {
            "ETH": "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",  # ETH placeholder
            "DOT": "0x7083609fce4d1d8dc0c979aab8c869ea2c873402",  # Polkadot (DOT) on Ethereum
        }
        
        # USD stable coins for price reference 
        self.usd_tokens = {
            "USDC": "0xA0b86a33E6441066CE6b8B0d5A5a57d4E47e0b12",  # USDC on mainnet
            "USDT": "0xdAC17F958D2ee523a2206206994597C13D831ec7",  # USDT on mainnet
        }
        
        if not self.api_key:
            logger.warning("1inch API key not provided. Price endpoints may be limited.")

    async def get_token_price_usd(self, token_symbol: str, chain_id: int = 1) -> Optional[Dict]:
        """
        Get token price in USD using 1inch Spot Price API
        
        Args:
            token_symbol: Token symbol (ETH, DOT)
            chain_id: Blockchain chain ID (default: 1 for Ethereum mainnet)
            
        Returns:
            Dict with price information or None if error
        """
        try:
            if token_symbol not in self.token_addresses:
                raise ValueError(f"Token {token_symbol} not supported. Supported tokens: {list(self.token_addresses.keys())}")
            
            token_address = self.token_addresses[token_symbol]
            usdc_address = self.usd_tokens["USDC"]
            
            # Use 1inch Spot Price API v1.1
            url = f"{self.base_url}/spot-price/v1.1/{chain_id}"
            
            headers = {
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json"
            }
            
            # Request body for spot price
            payload = {
                "src": token_address,
                "dst": usdc_address,
                "amount": "1000000000000000000" if token_symbol == "ETH" else "10000000000000000000",  # 1 token in wei
                "includeTokensInfo": True,
                "includeProtocols": False
            }
            
            async with httpx.AsyncClient() as client:
                response = await client.post(url, json=payload, headers=headers)
                
                if response.status_code == 200:
                    data = response.json()
                    
                    # Calculate price from the response
                    dst_amount = int(data.get("dstAmount", "0"))
                    if token_symbol == "ETH":
                        # ETH has 18 decimals, USDC has 6 decimals
                        price_usd = dst_amount / 1e6  # Convert from USDC smallest unit
                    else:  # DOT
                        # DOT has 10 decimals on Polkadot, but on Ethereum it typically has 10 decimals
                        # We're using 10 DOT (10 * 1e18) as amount
                        price_usd = (dst_amount / 1e6) / 10  # Convert from USDC and divide by 10 DOT
                    
                    return {
                        "token": token_symbol,
                        "price_usd": round(price_usd, 6),
                        "chain_id": chain_id,
                        "source": "1inch",
                        "timestamp": data.get("timestamp") or int(__import__("time").time())
                    }
                else:
                    logger.error(f"1inch API error for {token_symbol}: {response.status_code} - {response.text}")
                    return None
                    
        except Exception as e:
            logger.error(f"Error getting {token_symbol} price: {str(e)}")
            return None

    async def get_eth_price_usd(self, chain_id: int = 1) -> Optional[Dict]:
        """Get ETH price in USD"""
        return await self.get_token_price_usd("ETH", chain_id)

    async def get_polkadot_price_usd(self, chain_id: int = 1) -> Optional[Dict]:
        """Get Polkadot (DOT) price in USD"""
        return await self.get_token_price_usd("DOT", chain_id)

    async def get_multiple_prices(self, tokens: list, chain_id: int = 1) -> Dict:
        """Get prices for multiple tokens"""
        results = {}
        for token in tokens:
            price_data = await self.get_token_price_usd(token, chain_id)
            results[token] = price_data
        return results 

File: swap/src/services/priceService.ts
import axios from 'axios';
import { logger } from '../utils/logger';

interface TokenPriceResponse {
  token: string;
  price_usd: number;
  chain_id: number;
  source: string;
  timestamp: number;
}

// 1inch Price API response format
interface OneInchPriceResponse {
  [tokenAddress: string]: string; // price as string
}

export class PriceService {
  private apiKey: string;
  private baseUrl: string;
  
  // Token addresses for price queries
  private tokenAddresses = {
    'ETH': '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee', // ETH placeholder
    'DOT': '0x7083609fce4d1d8dc0c979aab8c869ea2c873402', // Polkadot (DOT) on Ethereum
  };

  constructor() {
    this.apiKey = process.env.ONEINCH_API_KEY || '';
    this.baseUrl = 'https://api.1inch.dev';
    
    // Debug logging to check environment variables
    logger.info('PriceService initialized', {
      hasApiKey: !!this.apiKey,
      apiKeyLength: this.apiKey.length,
      apiKeyStart: this.apiKey ? this.apiKey.substring(0, 8) + '...' : 'undefined'
    });
    
    if (!this.apiKey) {
      logger.warn('1inch API key not provided. Price endpoints may be limited.');
    }
  }

  async getTokenPriceUsd(tokenSymbol: string, chainId: number = 1): Promise<TokenPriceResponse | null> {
    try {
      if (!(tokenSymbol in this.tokenAddresses)) {
        throw new Error(`Token ${tokenSymbol} not supported. Supported tokens: ${Object.keys(this.tokenAddresses).join(', ')}`);
      }

      const tokenAddress = this.tokenAddresses[tokenSymbol as keyof typeof this.tokenAddresses];

      // Use 1inch Price API v1.1 with correct endpoint
      const url = `${this.baseUrl}/price/v1.1/${chainId}`;
      
      const headers = {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json'
      };

      // Request body for price API (correct format)
      const payload = {
        tokens: [tokenAddress],
        currency: 'USD'
      };

      logger.info(`Getting ${tokenSymbol} price from 1inch Price API`, { tokenAddress, chainId });

      const response = await axios.post<OneInchPriceResponse>(url, payload, { headers });

      if (response.status === 200) {
        const data = response.data;
        
        // Get price from the response (normalize address to lowercase for lookup)
        const priceString = data[tokenAddress.toLowerCase()];
        
        if (!priceString) {
          logger.error(`No price data returned for ${tokenSymbol} (${tokenAddress})`);
          return null;
        }

        const priceUsd = parseFloat(priceString);

        const result: TokenPriceResponse = {
          token: tokenSymbol,
          price_usd: Math.round(priceUsd * 1000000) / 1000000, // Round to 6 decimal places
          chain_id: chainId,
          source: '1inch',
          timestamp: Math.floor(Date.now() / 1000)
        };

        logger.info(`${tokenSymbol} price retrieved successfully`, result);
        return result;
      } else {
        logger.error(`1inch API error for ${tokenSymbol}:`, {
          status: response.status,
          data: response.data
        });
        return null;
      }

    } catch (error) {
      logger.error(`Error getting ${tokenSymbol} price:`, error);
      return null;
    }
  }

  async getEthPriceUsd(chainId: number = 1): Promise<TokenPriceResponse | null> {
    return this.getTokenPriceUsd('ETH', chainId);
  }

  async getPolkadotPriceUsd(chainId: number = 1): Promise<TokenPriceResponse | null> {
    return this.getTokenPriceUsd('DOT', chainId);
  }

  async getMultiplePrices(tokens: string[], chainId: number = 1): Promise<{ [token: string]: TokenPriceResponse | null }> {
    try {
      // Get all token addresses
      const tokenAddresses = tokens.map(token => {
        if (!(token in this.tokenAddresses)) {
          throw new Error(`Token ${token} not supported`);
        }
        return this.tokenAddresses[token as keyof typeof this.tokenAddresses];
      });

      const url = `${this.baseUrl}/price/v1.1/${chainId}`;
      
      const headers = {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json'
      };

      const payload = {
        tokens: tokenAddresses,
        currency: 'USD'
      };

      logger.info(`Getting multiple prices from 1inch Price API`, { tokens, chainId });

      const response = await axios.post<OneInchPriceResponse>(url, payload, { headers });

      const results: { [token: string]: TokenPriceResponse | null } = {};

      if (response.status === 200) {
        const data = response.data;

        tokens.forEach((token, index) => {
          const tokenAddress = tokenAddresses[index];
          const priceString = data[tokenAddress.toLowerCase()];
          
          if (priceString) {
            const priceUsd = parseFloat(priceString);
            results[token] = {
              token,
              price_usd: Math.round(priceUsd * 1000000) / 1000000,
              chain_id: chainId,
              source: '1inch',
              timestamp: Math.floor(Date.now() / 1000)
            };
          } else {
            results[token] = null;
          }
        });

        logger.info('Multiple prices retrieved successfully', results);
      } else {
        logger.error('1inch API error for multiple prices:', {
          status: response.status,
          data: response.data
        });
        
        // Set all results to null on error
        tokens.forEach(token => {
          results[token] = null;
        });
      }

      return results;
    } catch (error) {
      logger.error('Error getting multiple prices:', error);
      
      // Return null for all tokens on error
      const results: { [token: string]: TokenPriceResponse | null } = {};
      tokens.forEach(token => {
        results[token] = null;
      });
      return results;
    }
  }

  // Helper method to check if 1inch API is available
  async checkApiAvailability(): Promise<boolean> {
    try {
      const response = await axios.get(`${this.baseUrl}/healthcheck`, {
        headers: {
          'Authorization': `Bearer ${this.apiKey}`
        }
      });
      return response.status === 200;
    } catch (error) {
      logger.error('1inch API availability check failed:', error);
      return false;
    }
  }
} 